internal/consensus/state.go

```diff
@@ -2367,6 +2424,99 @@ func (cs *State) addProposalBlockPart(msg *BlockPartMessage, peerID p2p.ID) (add
 	return added, nil
 }
 
+// NOTE: blob is unvalidated bytes.
+// Asynchronously triggers either enterPrevote (before we timeout of propose) or tryFinalizeCommit,
+// once we have the full block. <- TODO: wrong
+func (cs *State) addProposalBlobPart(msg *BlobPartMessage, peerID p2p.ID) (added bool, err error) {
+	height, round, part := msg.Height, msg.Round, msg.Part
+
+	// Blobs might be reused, so round mismatch is OK
+	if cs.Height != height {
+		cs.Logger.Debug("Received blob part from wrong height", "height", height, "round", round)
+		// Todo: Implement metrics
+		//cs.metrics.BlobGossipPartsReceived.With("matches_current", "false").Add(1)
+		return false, nil
+	}
+
+	// We're not expecting a blob part.
+	if cs.ProposalBlobParts == nil {
+		// Todo: Implement metrics
+		//cs.metrics.BlobGossipPartsReceived.With("matches_current", "false").Add(1)
+		// NOTE: this can happen when we've gone to a higher round and
+		// then receive parts from the previous round - not necessarily a bad peer.
+		cs.Logger.Debug(
+			"Received a blob part when we are not expecting any",
+			"height", height,
+			"round", round,
+			"index", part.Index,
+			"peer", peerID,
+		)
+		return false, nil
+	}
+
+	added, err = cs.ProposalBlobParts.AddPart(part)
+	if err != nil {
+		// Todo: Implement metrics
+		//if errors.Is(err, types.ErrPartSetInvalidProof) || errors.Is(err, types.ErrPartSetUnexpectedIndex) {
+		//	cs.metrics.BlobGossipPartsReceived.With("matches_current", "false").Add(1)
+		//}
+		return added, err
+	}
+
+	// Todo: Implement metrics
+	//cs.metrics.BlobGossipPartsReceived.With("matches_current", "true").Add(1)
+	if !added {
+		// NOTE: we are disregarding possible duplicates above where heights dont match or we're not expecting blob parts yet
+		// but between the matches_current = true and false, we have all the info.
+		// Todo: Implement metrics
+		//cs.metrics.DuplicateBlobPart.Add(1)
+	} else {
+		cs.evsw.FireEvent(types.EventProposalBlobPart, msg)
+	}
+
+	count, total := cs.ProposalBlobParts.Count(), cs.ProposalBlobParts.Total()
+	cs.Logger.Debug("Receive blob part", "height", height, "round", round,
+		"index", part.Index, "count", count, "total", total, "from", peerID)
+
+	// Todo: Implement blob configuration
+	//maxBytes := cs.state.ConsensusParams.Blob.MaxBytes
+	//if maxBytes == -1 {
+	//	maxBytes = int64(types.MaxBlockSizeBytes)
+	//}
+	//if cs.ProposalBlockParts.ByteSize() > maxBytes {
+	//	return added, fmt.Errorf("total size of proposal block parts exceeds maximum block bytes (%d > %d)",
+	//		cs.ProposalBlockParts.ByteSize(), maxBytes,
+	//	)
+	//}
+	if added && cs.ProposalBlockParts.IsComplete() && cs.ProposalBlobParts.IsComplete() {
```

Comment:
Same comment as above for block parts. What if we have all the blob parts but not te block parts, we will essentially never assemble the blob. So here `cs.ProposalBlockParts.IsComplete()` could be moved to 2512. Or changed to a check whether `cs.ProposalBlock != nil`